lib/winapi/filemanip.c

```diff
@@ -241,6 +241,138 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
+        sizeof(networkOpenInformation),
+        FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    // Fail overwriting the file if the target == the source. This is
+    // intentionally done after the check for existence of the source
+    // to set the most appropriate LastError value.
+    //
+    // NOTE: FATX does not have ".." entries in directory tables, so it is not
+    // possible to overlap a file via "subdir/../file".
+    if (!strcmp(lpExistingFileName, lpNewFileName)) {
+      NtClose(sourceHandle);
+      SetLastError(ERROR_SHARING_VIOLATION);
+      return FALSE;
+    }
+
+    status = NtCreateFile(
+            &targetHandle,
+            FILE_GENERIC_WRITE,
+            &objectAttributes,
+            &ioStatusBlock,
+            &networkOpenInformation.AllocationSize,
+            networkOpenInformation.FileAttributes,
+            FILE_SHARE_READ | FILE_SHARE_WRITE,
+            bFailIfExists ? FILE_CREATE : FILE_SUPERSEDE,
+            FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT);
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    status = NtAllocateVirtualMemory(&readBuffer,
+                                      0,
+                                      &readBufferRegionSize,
+                                      MEM_RESERVE | MEM_COMMIT,
+                                      PAGE_READWRITE);
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        NtClose(targetHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    while (TRUE) {
+        const BYTE *bufferPos = readBuffer;
+        if (!ReadFile(sourceHandle, readBuffer, readBufferRegionSize, &bytesRead, NULL)) {
+            NtFreeVirtualMemory(&readBuffer, &readBufferRegionSize, MEM_RELEASE);
+            NtClose(sourceHandle);
+            NtClose(targetHandle);
+            return FALSE;
+        }
+
+        if (!bytesRead) {
+            break;
+        }
+
+        while (bytesRead > 0) {
+            DWORD bytesWritten = 0;
+            if (!WriteFile(targetHandle, bufferPos, bytesRead, &bytesWritten, NULL)) {
+                NtFreeVirtualMemory(&readBuffer, &readBufferRegionSize, MEM_RELEASE);
+                NtClose(sourceHandle);
+                NtClose(targetHandle);
+                return FALSE;
+            }
+            bytesRead -= bytesWritten;
+            bufferPos += bytesWritten;
+        }
+    }
+
+    NtFreeVirtualMemory(&readBuffer, &readBufferRegionSize, MEM_RELEASE);
+
```

Comment:
> I was under the impression that passing the file attributes to `NtCreateFile` above would already copy them over when creating/superseding the file.

That should be enough to copy the file attribute flags, correct.

> I also assumed that the access info should actually be based on the time of copy rather than brought over from the original.

The wording in the MSDN is a bit ambiguous, so I checked on Windows. It appears that only `LastWriteTime` is copied from the original file, the others are left as-is.

Reply:
Cool, looks like passing 0 should cause it to retain the copied file's value so I'm explicitly copying over `LastWriteTime` and (re)setting `FileAttributes`.